src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -29,26 +34,48 @@ public static partial class Base64
         /// </summary> 
         public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true)
         {
+            // PERF: use uint to avoid the sign-extensions
+            uint sourceIndex = 0;
+            uint destIndex = 0;
+
+            if (utf8.IsEmpty)
+                goto DoneExit;
+
             ref byte srcBytes = ref MemoryMarshal.GetReference(utf8);
             ref byte destBytes = ref MemoryMarshal.GetReference(bytes);
 
             int srcLength = utf8.Length & ~0x3;  // only decode input up to the closest multiple of 4.
             int destLength = bytes.Length;
+            int maxSrcLength = srcLength;
+            int decodedLength = GetMaxDecodedFromUtf8Length(srcLength);
 
-            int sourceIndex = 0;
-            int destIndex = 0;
+            // max. 2 padding chars
+            if (destLength < decodedLength - 2)
+            {
+                // For overflow see comment below
+                maxSrcLength = destLength / 3 * 4;
+            }
 
-            if (utf8.Length == 0)
-                goto DoneExit;
+            if (Avx2.IsSupported && maxSrcLength >= 45)
```

Comment:
Is the length here taking into account other potential drawbacks from executing 256-bit instructions?

Some examples are:
* For unaligned data, reads/writes crossing a cache-line boundary will happen twice as frequently (same with crossing page boundaries, for large enough data)
* Additional saving/restoring of the upper 128-bits across method call boundaries
* Additional `vzeroupper` calls
* Possible frequency downscaling when executing a "heavy" 256-bit workload (this one tends to make micro-benches look good, but real world scenarios can actually regress)

Reply:
The length here doesn't take any of your points / potential AVX drawbacks into account.

I thought about aligning writes for encoding / reads for decoding, as they are multiple of fours, and so (in theory) this could be done. But it's not easy to do this without "eating" up quite a lot in a scalar way.

Do you have a suggestion what to do with the length / what it should take specifically into account?

Reply:
> Do you have a suggestion what to do with the length / what it should take specifically into account?

The first three can probably just be profiled multiple times with varying ranges of input data. The last one is really hard to determine outside of profiling real-world scenarios.

> But it's not easy to do this without "eating" up quite a lot in a scalar way.

Did you look at processing the leading/trailing elements via vectorization as well (which should give you, at most, 2 unaligned read/writes)? I think you might have mentioned trying this on the other thread...

Reply:
> processing the leading/trailing elements via vectorization

You're correct that I tried this, but w/o success, as the bookkeeping for this produces more overhead than the scalar processing. 
Base64 is a bit tricky with vectorization, as not all elements of the vector produce usable results. Some elements are just 0, and will get overwritten in the next iteration / in the scalar remainder. Then there's the padding. To take all this into account produces quite a lot overhead -- or I'm not thinking clever enough to come with an easy and correct solution for this.